addons/costume-editor-shortcuts/userscript.js

```diff
@@ -0,0 +1,141 @@
+/**
+ * Scratch Addons Script to add Tool Shortcuts to Costume Editor
+ */
+export default async function ({ addon, global, console, msg }) {
+  const COSTUME_EDITOR_TAB_INDEX = 1;
+  const toolNameToShortcut = {
+    Select: "s",
+    Reshape: "a",
+    Brush: "b",
+    Eraser: "e",
+    Fill: "f",
+    Text: "t",
+    Line: "d",
+    Circle: "c",
+    Rectangle: "r",
+  };
+
+  /**
+   * Reverse the mapping above to lookup by shortcut key
+   */
+  const shortcutToToolName = Object.fromEntries(Object.entries(toolNameToShortcut).map(([key, value]) => [value, key]));
+
+  let isInitialized = false;
+  let isUserTyping = false;
+  let prevEditorTabIndex = 0;
+
+  /**
+   * Initialize redux and listen for when a new tab is active.
+   */
+  addon.tab.redux.initialize();
+  addon.tab.redux.addEventListener("statechanged", isNewTabActive);
+
+  /**
+   * If costume editor is open, initialize tool shortcuts, otherwise clean them up if needed.
+   */
+  async function isNewTabActive() {
+    const activeIndex = addon.tab.redux.state.scratchGui.editorTab.activeTabIndex;
+    if (prevEditorTabIndex != activeIndex) {
+      prevEditorTabIndex = activeIndex;
+      if (activeIndex === COSTUME_EDITOR_TAB_INDEX) {
+        await initialize();
+      } else if (activeIndex !== COSTUME_EDITOR_TAB_INDEX && isInitialized) {
+        await cleanup();
+      }
+    }
+  }
+
+  /**
+   * Setup keydown listeners for shortcuts and and update UI to display available shortcuts.
+   */
+  async function initialize() {
+    if (isInitialized) return; // Prevent double initialization
+    isInitialized = true;
+    document.addEventListener("keydown", handleKeyDown);
+    document.addEventListener("focusin", userStartedTyping);
+    document.addEventListener("focusout", userStoppedTyping);
```

Comment:
I haven't tested my suggestion but I think if you add the listeners to an element in the costume editor you won't have to clean them up.

Reply:
I tried this initially, but had issues getting anything to fire when the handlers were attached to any of the costume editor divs (no matter where I clicked). Probably user error, but `document` just worked.

Here's an example of what I tried:
```
  // var gui = await addon.tab.waitForElement("[class*='gui_tab-panel']");
  // var gui = await addon.tab.waitForElement("[class*='paint-editor_editor-container']");
  var gui = await addon.tab.waitForElement("#react-tabs-3");
  gui.addEventListener("keydown", handleKeyDown);
  gui.addEventListener("focusin", userStartedTyping);
  gui.addEventListener("focusout", userStoppedTyping);
```

I also wondered if there'd be an advantage to using document to not interrupt the workflow (when you're toggling between your game and costume editing). For example, if you eye droppered something on the game view or stage, your costume shortcuts would still work.